src/System.Text.Json/src/System/Text/Json/Serialization/JsonSnakeCaseNamingPolicy.cs

```diff
@@ -0,0 +1,72 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Globalization;
+
+namespace System.Text.Json
+{
+    internal sealed class JsonSnakeCaseNamingPolicy : JsonNamingPolicy
+    {
+        public override string ConvertName(string name)
+        {
+            if (string.IsNullOrEmpty(name))
+            {
+                return name;
+            }
+
+            const char underscore = '_';
+            const UnicodeCategory noneCategory = UnicodeCategory.Control;
+
+            StringBuilder builder = new StringBuilder();
```

Comment:
This is going to add a fair amount of allocation per conversion.

Reply:
Oops, a good catch. Is it okay to use the following capacity?

```csharp
var capacity = name.Length + Math.Min(2, name.Length / 5);
// 5 is a median word length in English
```